pkg/kwok/metrics/metrics.go

```diff
@@ -74,281 +71,484 @@ func NewMetricsUpdateHandler(conf UpdateHandlerConfig) (*UpdateHandler, error) {
-				}
-				return nil
+		keys := make([]string, 0, len(pods))
+		for _, pod := range pods {
+			data.Pod = pod.Pod
+			gauge, key, err := h.getOrRegisterGauge(metricConfig, data)
+			if err != nil {
+				return nil, err
 			}
 
-			eval, err := h.environment.Compile(m.Value)
+			result, err := eval.EvaluateFloat64(data)
 			if err != nil {
-				return fmt.Errorf("failed to compile metric %q: %w", metric.Name, err)
+				return nil, fmt.Errorf("failed to evaluate metric %q: %w", metricConfig.Name, err)
 			}
+			gauge.Set(result)
+			keys = append(keys, key)
+		}
+		return keys, nil
+	case internalversion.DimensionContainer:
+		pods, ok := h.getPodsInfo(nodeName)
+		if !ok {
+			logger := log.FromContext(ctx)
+			logger.Warn("pods not found", "node", nodeName)
+			return nil, nil
+		}
 
-			var once sync.Once
-			registered := false
-			updateFunc := func() error {
-				nodeInfo, ok := h.getNodeInfo(metric.Name)
-				if !ok {
-					if !registered {
-						return fmt.Errorf("failed to find node %q", metric.Name)
-					}
-					val.Set(0)
-					return nil
-				}
-
-				var err error
-				once.Do(func() {
-					err = registerFunc(nodeInfo.Node)
-					if err != nil {
-						registered = true
-					}
-				})
+		keys := make([]string, 0, len(pods))
+		for _, pod := range pods {
+			data.Pod = pod.Pod
+			for _, container := range pod.Pod.Spec.Containers {
+				container := container
+				data.Container = &container
+				gauge, key, err := h.getOrRegisterGauge(metricConfig, data)
 				if err != nil {
-					return err
+					return nil, err
 				}
-
-				result, err := eval.EvaluateFloat64(nodeInfo.Node)
+				result, err := eval.EvaluateFloat64(data)
 				if err != nil {
-					return fmt.Errorf("failed to evaluate metric %q: %w", m.Name, err)
+					return nil, fmt.Errorf("failed to evaluate metric %q: %w", metricConfig.Name, err)
 				}
-				val.Set(result)
-				return nil
+				gauge.Set(result)
+				keys = append(keys, key)
 			}
-			h.updates = append(h.updates, updateFunc)
-		case KindCounter:
-			counter := atomic.Int64{}
+		}
+		return keys, nil
+	default:
+		return nil, fmt.Errorf("unknown dimension %q", metricConfig.Dimension)
+	}
+}
 
-			registerFunc := func(node *corev1.Node) error {
-				labels, err := h.createLabels(preparedLabels, node)
-				if err != nil {
-					return fmt.Errorf("failed to evaluate labels: %w", err)
-				}
-				val := prometheus.NewCounterFunc(
-					prometheus.CounterOpts{
-						Name:        m.Name,
-						Help:        m.Help,
-						ConstLabels: labels,
-					},
-					func() float64 {
-						return float64(counter.Load())
-					},
-				)
-
-				err = h.registry.Register(val)
-				if err != nil {
-					return fmt.Errorf("failed to register counter %q: %w", m.Name, err)
-				}
-				return nil
-			}
+func (h *UpdateHandler) updateCounter(ctx context.Context, metricConfig *internalversion.MetricConfig, nodeName string) ([]string, error) {
+	eval, err := h.environment.Compile(metricConfig.Value)
```

Comment:
This all looks amazing, just one small comment I have here.
Since the updates are frequent, we can probably cache the compiled program. The program ast is the same, so we can get it from a cache based on the value. If two metrics are configuring the same value, the same eval can be used
And since metrics are configured, we don't need to think about invalidating the cache right? The number of programs is <= number of metricConfig values (In theory, two metrics may have the same value, so the same program can be used to evaluate them)
Again, this is just for performance, but I admire what you did here :smile: @wzshiming

Reply:
> Since the updates are frequent, we can probably cache the compiled program. The program ast is the same, so we can get it from a cache based on the value. If two metrics are configuring the same value, the same eval can be used

It has always been 😄

https://github.com/kubernetes-sigs/kwok/blob/876ac0fc4c14b7f55d2a47f4d06473605bb9879a/pkg/kwok/metrics/cel/evaluate.go#L148-L151

> And since metrics are configured, we don't need to think about invalidating the cache right? The number of programs is <= number of metricConfig values (In theory, two metrics may have the same value, so the same program can be used to evaluate them)

That makes sense, I will add the value of the cache.